File: Routing/consumer/src/main/scala/ru/example/rmq/routing/Consumer.scala
package ru.example.rmq.routing

import com.rabbitmq.client._

object Consumer {
  def main(args: Array[String]): Unit = {
    val EXCHANGE_NAME = "direct_logs"

    if (args.length < 1) {
      println("Usage: Consumer host [info] [warning] [error]")
      sys.exit(-1)
    }

    val host    = args(0)
    val factory = new ConnectionFactory
    factory.setHost(host)
    val connection = factory.newConnection
    val channel    = connection.createChannel

    channel.exchangeDeclare(EXCHANGE_NAME, BuiltinExchangeType.DIRECT)
    val queueName = channel.queueDeclare().getQueue

    args.slice(1, args.length).foreach { severity =>
      channel.queueBind(queueName, EXCHANGE_NAME, severity)
    }

    println("[*] Waiting for messages. To exit press CTRL+C")
    channel.basicConsume(
      queueName,
      true,
      (_: String, delivery: Delivery) => {
        val message = new String(delivery.getBody, "UTF-8")
        println(s"[x] Received '${delivery.getEnvelope.getRoutingKey}': '$message'")
      },
      (_: String, _: ShutdownSignalException) => {}
    )
  }
}

File: Tutorials/Java/4 Routing/Consumer/src/main/java/Consumer.java
import com.rabbitmq.client.*;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;

public class Consumer {
    private final static String EXCHANGE_NAME = "direct_logs";
    private final static String user = "user";
    private final static String password = "password";
    private final static String virtualHost = "/";

    public static void main(String[] argv) throws Exception {
        if (argv.length < 1) {
            System.err.println("Usage: Consumer host [info] [warning] [error]");
            System.exit(1);
        }

        String host = argv[0];

        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost(host);
        factory.setUsername(user);
        factory.setPassword(password);
        factory.setVirtualHost(virtualHost);

        Connection connection = factory.newConnection();
        Channel channel = connection.createChannel();

        channel.exchangeDeclare(EXCHANGE_NAME, BuiltinExchangeType.DIRECT);
        String queueName = channel.queueDeclare().getQueue();

        for (String severity : Arrays.copyOfRange(argv, 1, argv.length)) {
            channel.queueBind(queueName, EXCHANGE_NAME, severity);
        }
        System.out.println(" [*] Waiting for messages. To exit press CTRL+C");

        DeliverCallback deliverCallback = (consumerTag, delivery) -> {
            String message = new String(delivery.getBody(), StandardCharsets.UTF_8);
            System.out.println(" [x] Received '" + delivery.getEnvelope().getRoutingKey() + "':'" + message + "'");
        };
        channel.basicConsume(queueName, true, deliverCallback, consumerTag -> {});
    }
}